modules/core/aaa/auth.py

```diff
@@ -4944,42 +4944,59 @@ def s3_accessible_query(self, method, table, c=None, f=None):
     # -------------------------------------------------------------------------
     # S3 Variants of web2py Authorization Methods
     # -------------------------------------------------------------------------
-    def s3_has_membership(self, group_id=None, user_id=None, role=None):
+    def has_membership(self, group_id=None, user_id=None, role=None):
         """
             Checks if user is member of group_id or role
 
-            Extends Web2Py's requires_membership() to add new functionality:
+            Extends Web2Py's has_membership() to add:
+                - Respect for the deleted flag in auth_membership
                 - Custom Flash style
-                - Uses s3_has_role()
+
+            Args:
+                group_id: the group ID to check membership for
+                user_id: the user ID to check (defaults to current user)
+                role: the role name (alternative to group_id)
+
+            Returns:
+                True if the user is a member of the group, False otherwise
         """
 
         # Allow override
         if self.override:
             return True
 
+        # Determine user_id
+        if not user_id and self.user:
+            user_id = self.user.id
+
+        # Resolve group_id from role if needed
         group_id = group_id or self.id_group(role)
         try:
             group_id = int(group_id)
         except:
-            group_id = self.id_group(group_id) # interpret group_id as a role
-
-        has_role = self.s3_has_role(group_id)
-
+            group_id = self.id_group(group_id)  # interpret group_id as a role
+
+        # Check membership in database with deleted flag
+        r = False
```

Comment:
We should use `ret` rather than `r,` because in Eden, `r` is typically used for `CRUDRequest`s, so this could  cause confusion. But I can do that after merge.